### PR TITLE
CRAM index code/test refactoring.

### PR DESCRIPTION
**Pull request by cmnbroad**
Proposed refactoring of some existing CRAM indexing code and tests, both as prep for https://github.com/samtools/htsjdk/issues/433 and to better align the CRAM test classes with the code being tested; proposed removal of some unused code; some minor test bug fixes. No net new functionality.

Summary of changes:

- CRAMIndexer - Renamed to CRAMBAIIndexer to better reflect it's intent; all of the code is for generating BAI indices from a CRAM file (and some of the test code for this class is already in a file
called CRAMBAIIndexerTest).

- CRAIIndex - Removed two public methods (findLastAlignedEntry and getLeftMost) that were only referenced by their test code in CRAIIndexTest. (Are these useful to keep as public methods - htsjdk never uses .crai indices directly other than to convert them to .bai ?)

- SamIndexes - Propose to delete. The methods in this class are all for directly opening or discerning between .bai and .crai files/streams, but are not used by htsjdk to do this and are not used anywhere in htsjdk other than SamIndexesTest. Is it worth maintaining these as public methods ?

- SamIndexesTest - Propose to delete. Contains tests for the unused the methods in SamIndexes, with the exception of two tests that were using intermediate methods in SamIndexes to test CRAIIndex methods; those two tests have been moved to CRAIIndexTest and changed to call CRAIndex directly.

- CRAMComplianceTest - Removed some unnecessary/unused code (all tests remain) and renamed -> CRAMRoundTripTest, since all of the code is for testing round tripping between BAM and CRAM v2.1/v3.0.

- CramFileWriterTest - Renamed ->CRAMFileWriterTest to match the case-sense of the other files and the class it contains. Moved one general round-trip test into CRAMRoundTripTests.

- CRAMFileWriterWithIndexTest - Propose to delete. There are two tests in here; one was moved to CRAMFileIndexTest. The other one (testUnnecessaryIO) seems overly complicated and unnatural. I'm not sure what its testing for; is there a useful test here ?

### CODE REVIEW

**Comment by cmnbroad** [comment]
@vadimzalunin Can you take a look at this and let me know if there is anything here you think we should keep ? I don't want to throw out anything useful, just clean up before I integrate the .crai writing code.

**Comment by vadimzalunin** [comment]
@cmnbroad sounds good to me. 

CRAMComplianceTest - the goal was to test how htsjdk and htslib comply with each other. The CRAM files in the test have been generated with htslib. I don't mind renaming the class though.

CRAMFileWriterWithIndexTest.testUnnecessaryIO() - the goal was to ensure that only relevant regions of a CRAM file are read during search queries. For example the code might read too many bytes and then filter irrelevant records out. Normally this could be detected by bad performance on large files but this is difficult to achieve in tests. Moving/renaming the method is fine with me but I would prefer to keep the test unless there is a better way to test for unnecessary IO.

**Comment by droazen** [comment]
Vote to keep (and start actually using) `SamIndexes`

**Comment by cmnbroad** [comment]
Ok, I'll restore the I'm going to restore the testUnnecessaryIO test and revert the deletion of SamIndexes.

**Comment by cmnbroad** [comment]
I've folded in the various comments I've received; the updated summary of changes is:

- CRAMIndexer - Renamed to CRAMBAIIndexer to reflect its content (all of the code is for generating BAI indices), to align it with the existing CRAMBAIIndexerTest, and to distinguish it from the soon-to-come CRAMCRAIIndexer
- CRAIIndex/CRAIndexTest - Removed two public methods (findLastAlignedEntry and getLeftMost) that were only referenced by their test code in CRAIIndexTest
- CRAMComplianceTest - Removed some unnecessary/unused code (all tests remain)
- CramFileWriterTest - Renamed ->CRAMFileWriterTest to match the case-sense of the other files and the class it contains.
- CRAIEntry - minor code cleanup

File: src/java/htsjdk/samtools/cram/CRAIEntry.java [hunk 1]
```diff
@@ -63,13 +54,51 @@ public CRAIEntry(final String line) throws CRAIIndex.CRAIIndexException {
         }
     }
 
-    @Override
-    public String toString() {
-        return String.format("%d\t%d\t%d\t%d\t%d\t%d", sequenceId, alignmentStart, alignmentSpan,
+    /**
+     * Serializ the entry to a CRAI index stream.
```
Comment (droazen) [inline src/java/htsjdk/samtools/cram/CRAIEntry.java]: Serializ -> Serialize

File: src/tests/java/htsjdk/samtools/CRAMComplianceTest.java [hunk 2]
```diff
@@ -147,16 +136,15 @@ private void assertSameRecords(int majorVersion, SAMRecord record1, SAMRecord re
...
-                Assert.assertEquals(record2.getReadBases(), record1.getReadBases());
-        }
+        /**
+         * Known issue: CRAM v2.1 doesn't handle reads with missing bases correctly. This causes '*' bases to arise when reading CRAM.
+         * Skipping the base comparison asserts.
+         */
+        if (record1.getReadBases() == SAMRecord.NULL_SEQUENCE && majorVersion < CramVersions.CRAM_v3.major)
+            ;
+        else
+            Assert.assertEquals(record2.getReadBases(), record1.getReadBases());
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/CRAMComplianceTest.java]: Refactor to remove empty if clause, and add curly braces

File: src/tests/java/htsjdk/samtools/cram/CRAIIndexTest.java [hunk 3]
```diff
@@ -77,57 +93,81 @@ private boolean allFoundEntriesIntersectQueryInFind(final List<CRAIEntry> index,
         return foundCount > 0;
     }
 
+    @Test(expectedExceptions = NullPointerException.class)
+    public void testCraiRequiresDictionary() throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        final GZIPOutputStream gos = new GZIPOutputStream(baos);
+        gos.close();
+        final BufferedInputStream bis = new BufferedInputStream(new ByteArrayInputStream(baos.toByteArray()));
+        CRAIIndex.openCraiFileAsBaiStream(bis, null);
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/cram/CRAIIndexTest.java]: Close `bis`

File: src/tests/java/htsjdk/samtools/cram/CRAIIndexTest.java [hunk 4]
```diff
@@ -77,57 +93,81 @@ private boolean allFoundEntriesIntersectQueryInFind(final List<CRAIEntry> index,
...
+        final SeekableStream baiStream = CRAIIndex.openCraiFileAsBaiStream(new SeekableBufferedStream(new SeekableFileStream(file)), dictionary);
+        Assert.assertNotNull(baiStream);
+
+        final DiskBasedBAMFileIndex bamIndex = new DiskBasedBAMFileIndex(baiStream, dictionary);
+        final BAMFileSpan span = bamIndex.getSpanOverlapping(entry.sequenceId, entry.alignmentStart, entry.alignmentStart);
+        Assert.assertNotNull(span);
+        final long[] coordinateArray = span.toCoordinateArray();
+        Assert.assertEquals(coordinateArray.length, 2);
+        Assert.assertEquals(coordinateArray[0] >> 16, entry.containerStartOffset);
+        Assert.assertEquals(coordinateArray[1] & 0xFFFF, 1);
```
Comment (droazen) [inline src/tests/java/htsjdk/samtools/cram/CRAIIndexTest.java]: Looks like these two test cases share much of the same code -- can you extract a shared helper method?

**Comment by droazen** [comment]
Review complete -- back to @cmnbroad. Merge after addressing comments.

**Pull request merged**